File: packages/python/src/anyapi/_pagination.py
"""Cursor pagination for the sync and async clients (SPEC 2.5, 3.5).

A ``Paginator`` walks a paginated SKU one page at a time, yielding either the
flattened items (``__iter__``) or whole ``RunResult`` pages (``.pages()``).

Walk semantics (frozen):
    * Page 1 is ``run(slug, input, options)``.
    * If ``output.found`` is false (or data is None), stop.
    * Yield each item from ``items_field``. If ``nextCursor`` is a non-empty
      string, set ``input["cursor"] = nextCursor`` and repeat; a null/empty
      ``nextCursor`` ends the walk.
    * ``options["max_items"]`` caps the TOTAL items yielded across all pages.
      It is NOT sent as the wire ``max_items`` shaping param from the iterator;
      the iterator manages paging itself.

The emitter targets the module-level helpers :func:`paginate` (sync) and
:func:`apaginate` (async): each takes the bound client, slug, input dict, and
the items field name, and returns the matching paginator.
"""

from __future__ import annotations

import copy
from collections.abc import AsyncIterator, Callable, Coroutine, Iterator
from typing import (
    TYPE_CHECKING,
    Any,
    Generic,
    TypeVar,
    cast,
)

from ._transport import as_dict
from .types import OutputFound, RunResult

if TYPE_CHECKING:
    from ._async_client import AsyncAnyAPI
    from ._client import AnyAPI

__all__ = ["Paginator", "AsyncPaginator", "paginate", "apaginate"]

Item = TypeVar("Item")
Data = TypeVar("Data")

_NEXT_CURSOR = "nextCursor"
_CURSOR = "cursor"

# A run callable: (slug, input, options) -> RunResult (sync) or coroutine (async).
SyncRun = Callable[[str, "dict[str, Any]", "Any"], RunResult[Any]]
AsyncRun = Callable[
    [str, "dict[str, Any]", "Any"], Coroutine[Any, Any, RunResult[Any]]
]


def _field(data: object, key: str) -> object:
    """Read ``key`` from a dict page or a pydantic data model."""
    if not isinstance(data, dict):
        return getattr(data, key, None)
    return as_dict(cast("object", data)).get(key)


def _extract(
    result: RunResult[Any], items_field: str
) -> tuple[list[Any], str | None]:
    """Pull (items, next_cursor) from a page, or ([], None) when not found."""
    output = result.output
    if not isinstance(output, OutputFound) or output.data is None:
        return [], None
    data: object = output.data
    raw_items = _field(data, items_field)
    next_cursor = _field(data, _NEXT_CURSOR)
    items: list[Any] = (
        list(cast("list[Any]", raw_items))
        if isinstance(raw_items, (list, tuple))
        else []
    )
    cursor = (
        next_cursor
        if isinstance(next_cursor, str) and next_cursor
        else None
    )
    return items, cursor


def _max_items(options: object) -> int | None:
    cap = as_dict(options).get("max_items")
    return cap if isinstance(cap, int) else None


def _page_options(options: object) -> object:
    """Strip max_items before sending: the iterator paces itself (SPEC 2.5)."""
    if not isinstance(options, dict):
        return options
    source = as_dict(cast("object", options))
    if "max_items" not in source:
        return source or None
    stripped: dict[str, object] = dict(source)
    stripped.pop("max_items", None)
    return stripped or None


class Paginator(Generic[Item, Data]):
    """Sync cursor walk over a paginated SKU."""

    def __init__(
        self,
        run: SyncRun,
        slug: str,
        input: dict[str, Any],
        items_field: str,
        options: Any = None,
    ) -> None:
        self._run = run
        self._slug = slug
        self._input = copy.deepcopy(input)
        self._items_field = items_field
        self._options = options
        self._cap = _max_items(options)
        self._page_options = _page_options(options)

    def _walk(self) -> Iterator[RunResult[Any]]:
        cursor: str | None = None
        while True:
            page_input = dict(self._input)
            if cursor is not None:
                page_input[_CURSOR] = cursor
            result = self._run(self._slug, page_input, self._page_options)
            yield result
            _, cursor = _extract(result, self._items_field)
            if cursor is None:
                return

    def pages(self) -> Iterator[RunResult[Data]]:
        """Yield whole RunResult pages (read ``cost_usd`` per page)."""
        yield from self._walk()

    def __iter__(self) -> Iterator[Item]:
        count = 0
        for result in self._walk():
            items, _ = _extract(result, self._items_field)
            for item in items:
                if self._cap is not None and count >= self._cap:
                    return
                yield item
                count += 1
            if self._cap is not None and count >= self._cap:
                return


class AsyncPaginator(Generic[Item, Data]):
    """Async cursor walk over a paginated SKU."""

    def __init__(
        self,
        run: AsyncRun,
        slug: str,
        input: dict[str, Any],
        items_field: str,
        options: Any = None,
    ) -> None:
        self._run = run
        self._slug = slug
        self._input = copy.deepcopy(input)
        self._items_field = items_field
        self._options = options
        self._cap = _max_items(options)
        self._page_options = _page_options(options)

    async def _walk(self) -> AsyncIterator[RunResult[Any]]:
        cursor: str | None = None
        while True:
            page_input = dict(self._input)
            if cursor is not None:
                page_input[_CURSOR] = cursor
            result = await self._run(self._slug, page_input, self._page_options)
            yield result
            _, cursor = _extract(result, self._items_field)
            if cursor is None:
                return

    async def pages(self) -> AsyncIterator[RunResult[Data]]:
        """Yield whole RunResult pages (read ``cost_usd`` per page)."""
        async for result in self._walk():
            yield result

    async def __aiter__(self) -> AsyncIterator[Item]:
        count = 0
        async for result in self._walk():
            items, _ = _extract(result, self._items_field)
            for item in items:
                if self._cap is not None and count >= self._cap:
                    return
                yield item
                count += 1
            if self._cap is not None and count >= self._cap:
                return


def paginate(
    client: AnyAPI,
    slug: str,
    input: dict[str, Any],
    items_field: str,
    options: Any = None,
) -> Paginator[Any, Any]:
    """Build a sync Paginator bound to the client's run seam (emitter target)."""
    run: SyncRun = client._run  # pyright: ignore[reportPrivateUsage]
    return Paginator(run, slug, input, items_field, options)


def apaginate(
    client: AsyncAnyAPI,
    slug: str,
    input: dict[str, Any],
    items_field: str,
    options: Any = None,
) -> AsyncPaginator[Any, Any]:
    """Build an async Paginator bound to the client's run seam (emitter target)."""
    run: AsyncRun = client._arun  # pyright: ignore[reportPrivateUsage]
    return AsyncPaginator(run, slug, input, items_field, options)

File: packages/typescript/src/core/pagination.ts
// Handwritten runtime core: cursor pagination. See SPEC.md 2.5. The emitter targets
// `paginate(core, slug, input, itemsField, bare, options)`. Named exports only; zero deps.

import type { ClientCore } from "./client.js";
import type { BareRunResult, RequestOptions, RunResult } from "./types.js";

/**
 * An async-iterable of flattened items that also exposes whole pages. See SPEC 2.5. `Page`
 * is the page result type: `RunResult<Data>` for found-data SKUs, `BareRunResult<Data>` for
 * bare SKUs (SPEC 1.2 erratum).
 */
export interface Paginator<Item, Page> extends AsyncIterable<Item> {
  /** Iterate whole pages (each a run result) instead of flattened items. */
  pages(): AsyncIterable<Page>;
}

/** Read the page data object from either envelope, or null when a found-data page is empty. */
function pageData(result: { output: unknown }, bare: boolean): Record<string, unknown> | null {
  if (bare) {
    const out = result.output;
    return out !== null && typeof out === "object"
      ? (out as Record<string, unknown>)
      : null;
  }
  const env = result.output as { found?: boolean; data?: unknown };
  if (!env || env.found !== true || env.data === null || env.data === undefined) {
    return null;
  }
  return env.data as Record<string, unknown>;
}

/**
 * Walk a paginated SKU: run page 1, then follow `nextCursor` until it is null/empty.
 * `itemsField` is the array property on the page data holding the page items. `bare`
 * selects the envelope: found-data (`output.data`) or bare (`output` is the data).
 *
 * Walk semantics (SPEC 2.5):
 * - Stop when the page data is null (found:false / null data, or a non-object bare output).
 * - Yield each item from `itemsField`, then follow a non-empty string `nextCursor` by
 *   setting `input.cursor` and repeating.
 * - `options.maxItems` caps the TOTAL items yielded across all pages. It is NOT sent as
 *   the wire `max_items` shaping param (the iterator manages paging itself).
 * - `.pages()` yields each run result (including the last) and stops on null nextCursor.
 */
export function paginate<Item, Page extends RunResult<unknown> | BareRunResult<unknown>>(
  core: ClientCore,
  slug: string,
  input: Record<string, unknown>,
  itemsField: string,
  bare: boolean,
  options?: RequestOptions,
): Paginator<Item, Page> {
  // The iterator manages max_items itself; never forward it to the wire.
  const wireOptions: RequestOptions | undefined = stripMaxItems(options);
  const maxItems = options?.maxItems;

  async function* walkPages(): AsyncGenerator<Page, void, unknown> {
    let cursor: string | undefined;
    for (;;) {
      const pageInput: Record<string, unknown> = { ...input };
      if (cursor !== undefined) {
        pageInput["cursor"] = cursor;
      }
      const result = (await core.run<unknown>(slug, pageInput, wireOptions)) as Page;
      yield result;

      const data = pageData(result, bare);
      if (data === null) {
        return;
      }
      const next = data["nextCursor"];
      if (typeof next !== "string" || next === "") {
        return;
      }
      cursor = next;
    }
  }

  async function* walkItems(): AsyncGenerator<Item, void, unknown> {
    let yielded = 0;
    for await (const page of walkPages()) {
      const data = pageData(page, bare);
      if (data === null) {
        return;
      }
      const items = data[itemsField];
      if (!Array.isArray(items)) {
        continue;
      }
      for (const item of items) {
        if (maxItems !== undefined && yielded >= maxItems) {
          return;
        }
        yield item as Item;
        yielded += 1;
      }
      if (maxItems !== undefined && yielded >= maxItems) {
        return;
      }
    }
  }

  const paginator: Paginator<Item, Page> = {
    [Symbol.asyncIterator]: () => walkItems(),
    pages: () => ({ [Symbol.asyncIterator]: () => walkPages() }),
  };
  return paginator;
}

/** Return options without the maxItems key (iterator caps totals client-side). */
function stripMaxItems(options?: RequestOptions): RequestOptions | undefined {
  if (!options) {
    return undefined;
  }
  const { maxItems: _maxItems, ...rest } = options;
  void _maxItems;
  return rest;
}
